### PR TITLE
🎉 New Source MarketingCloud / ExactTarget  (singer-based)

### PR DESCRIPTION
**Pull request by schlattk**
## What
Adding a sf-marketingcloud exacttarget source based on singer-source https://github.com/singer-io/tap-exacttarget

## How
*Describe the solution*
made a small adjustment in the dockerfile so this source excludes data extensions to solve a problem with the user interface.
## Recommended reading order
1. `x.java`
2. `y.python`

## Pre-merge Checklist
Expand the checklist which is relevant for this PR. 

<details><summary> <strong> Connector checklist </strong> </summary>
<p>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Secrets are annotated with `airbyte_secret` in output spec
- [ ] Unit & integration tests added as appropriate (and are passing)
    * Community members: please provide proof of this succeeding locally e.g: screenshot or copy-paste acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] `/test connector=connectors/<name>` command as documented [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-an-existing-connector) is passing. 
    * Community members can skip this, Airbyters will run this for you. 
- [ ] Code reviews completed
- [ ] Credentials added to Github CI if needed and not already present. [instructions for injecting secrets into CI](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#using-credentials-in-ci). 
- [ ] Documentation updated 
    - [ ] README
    - [ ] CHANGELOG.md
    - [ ] Reference docs in the `docs/integrations/` directory.
    - [ ] Build status added to [build page](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/builds.md)
- [ ] Build is successful
- [ ] Connector version bumped like described [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-a-connector)
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-a-connector)
- [ ] No major blockers
- [ ] PR merged into master branch
- [ ] Follow up tickets have been created
- [ ] Associated tickets have been closed & stakeholders notified
</p>
</details>

<details><summary> <strong> Connector Generator checklist </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

**Comment by sherifnada** [comment]
@schlattk thanks for the contribution! apologies for the late response, this must have fell through the cracks. Will review this soon and get back to you.

File: airbyte-integrations/connectors/source-sf-marketingcloud-singer/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+# Bash is installed for more convenient debugging.
+RUN apt-get update -y && apt-get install -y bash && apt-get install -y gcc && apt-get install -y git && rm -rf /var/lib/apt/lists/*
+
+WORKDIR /airbyte/integration_code
+COPY source_sf_marketingcloud_singer ./source_sf_marketingcloud_singer
+COPY main.py ./
+COPY setup.py ./
+RUN pip install .
+# this line can be removed if data extensions are required
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-sf-marketingcloud-singer/Dockerfile:11]: what is the upside to removing this? I think it makes me a little nervous to edit code via sed.

File: airbyte-integrations/connectors/source-sf-marketingcloud-singer/setup.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+# SOFTWARE.
+#
+
+from setuptools import find_packages, setup
+
+MAIN_REQUIREMENTS = [
+    "airbyte-cdk",
+    "Salesforce-FuelSDK",
+    "suds-py3",
+    # "tap-exacttarget @ git+https://github.com/schlattk/exacttarget-singer-clone",
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-sf-marketingcloud-singer/setup.py:31]: can this commment be removed?

File: airbyte-integrations/connectors/source-sf-marketingcloud-singer/source_sf_marketingcloud_singer/source.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+            transport = HttpAuthenticated(timeout=int(local_config.get("request_timeout", 900)))
+            auth_stub.soap_client.set_options(transport=transport)
+            return AirbyteConnectionStatus(status=Status.SUCCEEDED)
+        except Exception as e:
+            logger.info("Failed to auth using V2 endpoint")
+            raise e
+            logger.info("Login succeeded")
+            return AirbyteConnectionStatus(status=Status.FAILED, message=f"An exception occurred: {str(e)}")
+
+    # def discover_cmd(self, logger: AirbyteLogger, config_path: str) -> str:
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-sf-marketingcloud-singer/source_sf_marketingcloud_singer/source.py:91]: we can remove the commented out portion

File: buildSrc/src/main/groovy/airbyte-python.gradle [hunk 4]
```diff
@@ -16,6 +16,7 @@ class AirbytePythonPlugin implements Plugin<Project> {
         project.plugins.apply 'ru.vyarus.use-python'
 
         project.python {
+            pythonBinary = 'python'
```
Comment (sherifnada) [inline buildSrc/src/main/groovy/airbyte-python.gradle:19]: is this change necessary?

**Review by sherifnada — COMMENTED** [review]
@schlattk thanks for the PR! Just a few questions. 

Also, we'd love to merge this into Airbyte but it's going to take us some time to setup an ExactTarget sandbox account against which we can run integration tests. In the meantime, could you provide evidence that [acceptance tests](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector/testing-connectors) are succeeding for this connector?  . Here is an example of how configure it: 

* [test configuration](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-harvest/acceptance-test-config.yml)
* [bash runner](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-harvest/acceptance-test-docker.sh)

File: airbyte-integrations/connectors/source-sf-marketingcloud-singer/Dockerfile (hunk 1 above)
Comment (schlattk) [reply airbyte-integrations/connectors/source-sf-marketingcloud-singer/Dockerfile:11]: the problem is that including data extensions breaks the Airbyte UI - at least for our exacttarget account. The alternative to using sed is to create a clone of tap-exacttarget, modify it and refer to the clone as the dependency. We thought that using the one liner in the Dockerfile could be more easily removed by others who wanted to get the data extensions.

File: buildSrc/src/main/groovy/airbyte-python.gradle (hunk 4 above)
Comment (schlattk) [reply buildSrc/src/main/groovy/airbyte-python.gradle:19]: I'll remove it

**Comment by schlattk** [comment]
> @schlattk thanks for the PR! Just a few questions.
> 
> Also, we'd love to merge this into Airbyte but it's going to take us some time to setup an ExactTarget sandbox account against which we can run integration tests. In the meantime, could you provide evidence that [acceptance tests](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector/testing-connectors) are succeeding for this connector? . Here is an example of how configure it:
> 
> * [test configuration](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-harvest/acceptance-test-config.yml)
> * [bash runner](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-harvest/acceptance-test-docker.sh)

Do you have a singer source I can model the tests on? I have looked at a few and couldn't find any tests there. Just out of interest how long would you think it would take to until we can merge this?

File: airbyte-integrations/connectors/source-sf-marketingcloud-singer/Dockerfile (hunk 1 above)
Comment (michael4tasman) [reply airbyte-integrations/connectors/source-sf-marketingcloud-singer/Dockerfile:11]: @sherifnada , to expand a bit on @schlattk 's point, there is currently a bug in the Airbyte UI related to very large catalogs. If the data extensions from the Singer tap are included, it triggers this bug. When the bug in the UI is fixed, then the data extensions can be put back at that point.

File: airbyte-integrations/connectors/source-sf-marketingcloud-singer/setup.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+
+from setuptools import find_packages, setup
+
+MAIN_REQUIREMENTS = [
+    "airbyte-cdk",
```
Comment (keu) [inline airbyte-integrations/connectors/source-sf-marketingcloud-singer/setup.py:28]: please bind to a specific version or at least to compatible with, i.e. ~=0.1.. etc

File: airbyte-integrations/connectors/source-sf-marketingcloud-singer/setup.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    "Salesforce-FuelSDK",
+    "suds-py3",
+    "tap-exacttarget",
```
Comment (keu) [inline airbyte-integrations/connectors/source-sf-marketingcloud-singer/setup.py:31]: especially taking into account that we `sed` source file of this package

File: airbyte-integrations/connectors/source-sf-marketingcloud-singer/build.gradle [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+    id 'airbyte-source-acceptance-test'
+}
+
+airbytePython {
+    moduleDirectory 'source_sf_marketingcloud_singer'
+}
+
+dependencies {
+    implementation files(project(':airbyte-integrations:bases:source-acceptance-test').airbyteDocker.outputs)
+    implementation files(project(':airbyte-integrations:bases:base-singer').airbyteDocker.outputs)
```
Comment (keu) [inline airbyte-integrations/connectors/source-sf-marketingcloud-singer/build.gradle:13]: I don't think we need this line anymore since we using python3.7-slim as a base image

File: airbyte-integrations/connectors/source-sf-marketingcloud-singer/Dockerfile (hunk 1 above)
Comment (sherifnada) [reply airbyte-integrations/connectors/source-sf-marketingcloud-singer/Dockerfile:11]: Understood. In that case what we should do is override the `discover` method in source.py and remove the data extension streams. Something like : 

```

def discover(self, logger, config) -> AirbyteCatalog: 
    catalog = super().discover(logger, config)
    catalog.streams = [s for s in catalog.streams if s.name != "ABCXYZ"]
    return catalog
```

this doesn't alter the singer tap itself but just modifies the Airbyte connector shim to exclude that

File: airbyte-integrations/connectors/source-sf-marketingcloud-singer/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+RUN sed -i '/^    DataExtensionDataAccessObject,$/d' /usr/local/lib/python3.7/site-packages/tap_exacttarget/__init__.py
+
+
+ENV AIRBYTE_ENTRYPOINT "python /airbyte/integration_code/main.py"
+ENTRYPOINT ["python", "/airbyte/integration_code/main.py"]
+
+LABEL io.airbyte.version=0.1.0
+LABEL io.airbyte.name=airbyte/source-sf-marketingcloud
```
Comment (keu) [inline airbyte-integrations/connectors/source-sf-marketingcloud-singer/Dockerfile:19]:
```suggestion
LABEL io.airbyte.name=airbyte/source-sf-marketingcloud-singer
```

we reserve `airbyte/source-sf-marketingcloud` name for the native version of this connector

**Comment by keu** [comment]
@schlattk please check this branch https://github.com/airbytehq/airbyte/tree/keu/mc_singer_source
for the example of files that you need to have to enable SAT. Please follow TODOs trace to fix missing parts.
Tip  - you can use the output from discovery command to generate configured catalog file. Note: json_schema there doesn't need to be filled - can be empty `{}`

**Comment by keu** [comment]
small comment regarding naming, currently we have `source-salesforce-singer` connector, for consistency I think the easiest is to rename this one to `source-salesforce-marketingcloud-singer`, WDYT?

File: airbyte-integrations/connectors/source-sf-marketingcloud-singer/Dockerfile (hunk 1 above)
Comment (schlattk) [reply airbyte-integrations/connectors/source-sf-marketingcloud-singer/Dockerfile:11]: Hi sherif I'm having some problems implementing this as you suggest. currently we have this as our discover method:
`def discover_cmd(self, logger: AirbyteLogger, config_path: str) -> AirbyteCatalog:
        return f"tap-exacttarget --config {config_path} --discover"`
